storage, ui: add metrics for raft messages

rjnn opened the pull request:
Add additional metrics for raft messages to help in debugging: total
messages sent and received, transport queue length, dropped messages,
and ticks. Expose these metrics in the Admin UI, under the Nodes -> "Advanced
Internals" section. Closes #8645. cc @mberhault.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8803)

<!-- Reviewable:end -->

maxlang commented on ui/app/containers/nodeGraphs.tsx:
```diff
@@ -218,6 +218,46 @@ export default class extends React.Component<IInjectedProps, {}> {
+                <Metric name="cr.store.raft-messages.rcvd.MsgTransferLeader" title="MsgTransferLeader" nonNegativeRate />
+                <Metric name="cr.store.raft-messages.rcvd.MsgTimeoutNow" title="MsgTimeoutNow" nonNegativeRate />
+              </Axis>
+            </StackedAreaGraph>
```
It might make sense to break some of these out into different graphs? This could be hard to follow, especially since stacked area charts are already hard to read.

maxlang commented:
LGTM

Add some screenshots!

mberhault commented on storage/replica.go:
```diff
@@ -1872,6 +1872,16 @@ func (r *Replica) sendRaftMessage(msg raftpb.Message) {
+	for _, queue := range transportQueues {
+		queuedMsgs += int64(len(queue))
+	}
+	r.store.metrics.RaftTransportQueueSize.Update(queuedMsgs)
```
the metric update should go outside the lock. they have their own synchronization

mberhault commented on storage/store_metrics.go:
```diff
@@ -141,6 +228,34 @@ type StoreMetrics struct {
+	// metric references must exist as AddMetricStruct adds them by reflection
+	// on this struct and does not process map types.
+	// TODO(arjun): eliminate this duplication.
+	RaftRcvdMessages       map[raftpb.MessageType]*metric.Counter
```
let's at least make it private (s/RaftRcvdMessages/raftRcvdMessage/). `AddMetricStruct` will ignore it due to be inaccessible which is a bit clearer than being the wrong type.

mberhault commented on storage/store_metrics.go:
```diff
@@ -74,9 +75,95 @@ var (
+	// Raft message metrics.
+	metaRaftRcvdMessages = map[raftpb.MessageType]metric.Metadata{
+		raftpb.MsgHup: {
+			Name: "raft-messages.rcvd.MsgHup",
```
metric names get prometheus-ized to: `[a-zA-Z_:][a-zA-Z0-9_:]*` by replacing anything that doesn't match the regexp with `_`. This means that this name would get turned into: `raft_messages_rcvd_MsgHup`. Not great.

I'd also like to start normalizing the metrics names, at least for new ones. Existing ones will be renamed when refactors become easier (after the code yellow).

So for all of these, I propose:
`raft.received.hup`

tamird commented:
Reviewed 2 of 5 files at r1.
Review status: 2 of 5 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/store_metrics.go, line 252 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_u-r1-252:-KPy2eNrbOZd9zDLkfTm:bbiroab) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/storage/store_metrics.go#L252)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> let's at least make it private (s/RaftRcvdMessages/raftRcvdMessage/). `AddMetricStruct` will ignore it due to be inaccessible which is a bit clearer than being the wrong type.
</details>
Can't you make AddMetricStruct traverse into maps?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

mberhault commented on storage/store_metrics.go:
```diff
@@ -74,9 +75,95 @@ var (
+		Help: "Total number of outgoing messages from this store that were dropped by the transport",
+	}
+
+	metaRaftTicks = metric.Metadata{Name: "raft.ticks",
```
nit: put `Name` on a separate line as you do above.

mberhault commented on storage/store_metrics.go:
```diff
@@ -74,9 +75,95 @@ var (
+	}
+
+	metaRaftSentMessages = metric.Metadata{
+		Name: "raft-messages.sent.total",
```
ditto for the names here. I would go with `raft.sent.total`

mberhault commented on storage/store_metrics.go:
```diff
@@ -74,9 +75,95 @@ var (
+	metaRaftTicks = metric.Metadata{Name: "raft.ticks",
+		Help: "Total number of Raft ticks processed"}
+
+	metaRaftTransportQueueSize = metric.Metadata{Name: "raft.transportqueue.size",
```
is this for messages being sent only? If so, I would go with `raft.sent.pending`. Otherwise, maybe plain `raft.pending`.
Putting `size` in the name makes it sounds like capacity rather than current usage.

mberhault commented on storage/store_metrics.go:
```diff
@@ -206,8 +321,48 @@ func newStoreMetrics() *StoreMetrics {
+		RaftDroppedMessages:    metric.NewCounter(metaRaftDroppedMessages),
+		RaftTicks:              metric.NewCounter(metaRaftTicks),
+		RaftTransportQueueSize: metric.NewGauge(metaRaftTransportQueueSize),
+		RaftSnapshotCount:      metric.NewCounter(metaRaftSnapshotCount),
```
this isn't being set anywhere.

mberhault commented:
Review status: 2 of 5 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/store_metrics.go, line 252 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_u-r1-252:-KPy4LFPkmHq6CLB20nf:bhh759p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/storage/store_metrics.go#L252)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Can't you make AddMetricStruct traverse into maps?
</details>
we could, but before trying harder to examine data types, I want to do the refactor that limits the structs passed to stuff that is (or contains) metrics. That way, we can be sure we don't miss any (eg: by dying on unknown fields)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Leaving this one to @mberhault since he's been working on and caring about metrics.

mberhault commented:
I'm questioning the need for per-message metrics in the admin UI. This is going to be an ungrokable amount of data for a single graph.
I would propose instead doing a single count of all messages for the built-in tsdb and restricting the per-message type metrics to the exporter.
We can do the latter with the (soon to be out for review) labeled metrics (eg: you'll have one metric, but you'll say `raftMessages.Inc(<message type>, 1)`.

arjunravinarayan commented:
Most of the metrics are zero, but occasionally you do want to see if there is a nonzero count for some message type - adding a bunch of graphs that flatline isn't that great either. Take a look at the screenshots - this is what it looks like. If you still feel that way I can remove them.

![Screenshot 2016-08-24 15.38.19.png](https://files.reviewableusercontent.io/images/reviewable/1565833/mus4v2jJsUJGhzfXNahv7S9V/Screenshot%202016-08-24%2015.38.19.png)
![Screenshot 2016-08-24 15.38.28.png](https://files.reviewableusercontent.io/images/reviewable/1565833/y9-b6DHbraRc8bu5bCVyBwsm/Screenshot%202016-08-24%2015.38.28.png)

---

Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1881 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_s-r1-1881:-KPy224cXnzkUFVKMv4V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/storage/replica.go#L1881)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> the metric update should go outside the lock. they have their own synchronization
</details>
Done.

---

*[storage/store_metrics.go, line 88 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_u-r1-88:-KPy5mehDskdI0uEjv5G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/storage/store_metrics.go#L88)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> metric names get prometheus-ized to: `[a-zA-Z_:][a-zA-Z0-9_:]*` by replacing anything that doesn't match the regexp with `_`. This means that this name would get turned into: `raft_messages_rcvd_MsgHup`. Not great.
> 
> 
> 
> I'd also like to start normalizing the metrics names, at least for new ones. Existing ones will be renamed when refactors become easier (after the code yellow).
> 
> 
> 
> So for all of these, I propose:
> 
> `raft.received.hup` and derivations. I'm intentionally dropping `Msg`
</details>
Done.

---

*[storage/store_metrics.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_u-r1-150:-KPy5fcVWtgnCCu4TTA5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/storage/store_metrics.go#L150)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> ditto for the names here. I would go with `raft.sent.total`
</details>
Done.

---

*[storage/store_metrics.go, line 159 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_u-r1-159:-KPy38ps8XYQ_f42DF3W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/storage/store_metrics.go#L159)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> nit: put `Name` on a separate line as you do above.
</details>
Done.

---

*[storage/store_metrics.go, line 162 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_u-r1-162:-KPy5bCwW6yYWiMgdSMd:byvyago) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/storage/store_metrics.go#L162)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> is this for messages being sent only? If so, I would go with `raft.sent.pending`. Otherwise, maybe plain `raft.pending`.
> 
> Putting `size` in the name makes it sounds like capacity rather than current usage.
</details>
Yes. Done.

---

*[storage/store_metrics.go, line 252 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_u-r1-252:-KPy24v_L5efeD6EFvik:bz274w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/storage/store_metrics.go#L252)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> we could, but before trying harder to examine data types, I want to do the refactor that limits the structs passed to stuff that is (or contains) metrics. That way, we can be sure we don't miss any (eg: by dying on unknown fields)
</details>
I'm with @mberhault here - I intentionally avoided messing with anything tricky like refactoring `AddMetricStruct`.

---

*[storage/store_metrics.go, line 347 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_u-r1-347:-KPy5lVY_vSGMmEyWqGC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/storage/store_metrics.go#L347)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> this isn't being set anywhere.
</details>
Done.

---

*[ui/app/containers/nodeGraphs.tsx, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_v-r1-240:-KPy3DNB494PVOT49Tn5:b-txeaiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/ui/app/containers/nodeGraphs.tsx#L240)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> It might make sense to break some of these out into different graphs? This could be hard to follow, especially since stacked area charts are already hard to read.
</details>
Take a look at the screenshot - I thought about it, but it's pretty readable as-is, because most of the values are zeros in the usual case (but when they are non-zero, it's important to have them be visible somewhere).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-:-KPy2uGfqu2UCpdOCtKw:b1u9x8i)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I would certainly look at that message type graph.

maxlang commented:
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[ui/app/containers/nodeGraphs.tsx, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_v-r1-240:-KPyKWlRtl6tDWFe006y:b-mtqebo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6240fabee508f8e05d5d1524ae864d21e4e88008/ui/app/containers/nodeGraphs.tsx#L240)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Take a look at the screenshot - I thought about it, but it's pretty readable as-is, because most of the values are zeros in the usual case (but when they are non-zero, it's important to have them be visible somewhere).
</details>
It's probably fine for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Don't feel pressured to put this in here, but I think there's one more measurement which is pretty useful, namely whether the lease holder [if any] is colocated with the Raft leader. I'm not sure what exactly we want to measure - certainly we want to know if that happens to many ranges (since that gives us an idea of how well our colocation mechanisms work in practice), but we also want to know if it happens to many proposals (in which case it could happen on a few hot ranges, and hot ranges are more likely to be in that state, too).

LMK what you think, if you'd rather get this in first I understand and we can file a new issue (after perhaps a short deliberation).

tschottdorf commented:
Actually, I think your `MsgProp` count does already count the second option in my last post (`MsgProp` should be the message for raft-internal proposal relays), except it also catches those emanating from reproposals, which isn't necessarily a bad thing).

Reproposals are another thing we should be counting. Perhaps we're doing ok, given that we are tracking the sum of two numbers we're not tracking.

mberhault commented on storage/store_metrics.go:
```diff
@@ -74,9 +75,91 @@ var (
-	metaRaftWorkingDurationNanos = metric.Metadata{Name: "process-raft.workingnanos"}
-	metaRaftTickingDurationNanos = metric.Metadata{Name: "process-raft.tickingnanos"}
+	metaRaftSelectDurationNanos = metric.Metadata{Name: "process-raft.waitingnanos",
+		Help: "Nanoseconds spent in store.processRaft() waiting"}
```
optional: if you feel like it, could you rename the `process-raft.*nanos` metrics to `raft.*.ns`
you'd have to adjust the UI as well.

mberhault commented on storage/store.go:
```diff
@@ -2575,6 +2578,9 @@ func (s *Store) processRaft() {
 					}
 				}
+
+				s.metrics.RaftTicks.Inc(1)
```
this can be done outside the `s.mu` lock

mberhault commented:
LGTM

Commit:
storage, ui: add metrics for Raft messages

Add additional metrics for Raft messages to help in debugging: total
messages sent and received, transport queue length, dropped messages,
and ticks. Expose these metrics in the Admin UI, under the "Advanced
Internals" section. Closes #8645.

arjunravinarayan commented:
Added another graph for counting leaders, leaseholders, and non-collocated leaseholders. PTAL @andreimatei @tschottdorf.

---

Review status: 0 of 5 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2582 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_t-r2-2582:-KQ1nOUDrgMZaGvTHvZu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd801b18ed59d888c713184192b8e6086b6617d2/storage/store.go#L2582)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> this can be done outside the `s.mu` lock
</details>
Done.

---

*[storage/store_metrics.go, line 79 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KPxyayZBYrdggCncH_u-r2-79:-KQ1nLwzFLN_TSHc-yUI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd801b18ed59d888c713184192b8e6086b6617d2/storage/store_metrics.go#L79)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> optional: if you feel like it, could you rename the `process-raft.*nanos` metrics to `raft.*.ns`
> 
> you'd have to adjust the UI as well.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-:-KQ2Js-CpzAzq24ahBjw:b39pinn)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1563 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4nAVDuqPbRvKl5yyP:-KQ4nAVEnU4ZgcIE38wa:bur59jz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1563)):*
> ```Go
> 	leaderID := r.mu.leaderID
> 
> 	if leaderID == r.mu.replicaID {
> ```

Doing this in `handleRaftReady` will skew the results, since replicas become ready in proportion to their activity. We want to do this per tick (or less often). 

---

*[storage/store_metrics.go, line 86 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4oMOHSmMxzBSmdUQ0:-KQ4oMOHSmMxzBSmdUQ1:bomsze5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L86)):*
> ```Go
> 
> 	// Raft message metrics.
> 	metaRaftReceivedHup = metric.Metadata{
> ```

The messages defined as "local" in https://github.com/coreos/etcd/blob/417b9469aa0f3708829f80dfc3325a576ed43ab0/raft/util.go#L49 never pass through the transfer so we'll never have non-zero values for them. These could be excluded from the metrics. 

---

*[storage/store_metrics.go, line 272 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4oknM-lR1vjvupioS:-KQ4oknM-lR1vjvupioT:b-sn1wwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L272)):*
> ```Go
> 	// These three gauges need to be atomically updated, but the values are aggregated from individual replicas using the private values.
> 	RaftLeaders                       *metric.Gauge
> 	raftLeaders                       *metric.Counter
> ```

This is a weird pattern. Why is the accumulator a `*metric.Counter` instead of a `*int64`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1539 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6W-gN1PNWhiaSsIVj:-KQ6W-gN1PNWhiaSsIVk:bbq862f) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1539)):*
> ```Go
> }
> 
> func (r *Replica) isLeaseHolder() bool {
> ```

cc @andreimatei you might find this useful in your PR which has that awkward variable name.

---

*[storage/replica.go, line 1563 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4nAVDuqPbRvKl5yyP:-KQ6XJzfsgZXheIF8LYm:b-lpxlno) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1563)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Doing this in `handleRaftReady` will skew the results, since replicas become ready in proportion to their activity. We want to do this per tick (or less often). 
</details>
The counter is being reset just before the iteration over replicas + call to `handleRaftReady` - what skew are you referring to?

---

*[storage/replica.go, line 1902 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6WO5LUIsPN9nBMQ1b:-KQ6WO5LUIsPN9nBMQ1c:bk0iod5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1902)):*
> ```Go
> 
> 	r.store.ctx.Transport.mu.Lock()
> 	transportQueues := r.store.ctx.Transport.mu.queues[false]
> ```

don't you want to grab all the queues?

```
var queuedMsgs int64
for _, transportQueues := range r.store.ctx.Transport.mu.queues {
  for _, queue := range transportQueues {
    queuedMsgs += int64(len(queue))
  }
}
```

---

*[storage/replica.go, line 1909 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6WjU6qbOo9uKSRnsd:-KQ6WjU7d5xK0c3jJpK0:bek2wun) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1909)):*
> ```Go
> 	r.store.ctx.Transport.mu.Unlock()
> 	r.store.metrics.RaftSentPending.Update(queuedMsgs)
> 	r.store.metrics.RaftSentMessages.Inc(1)
> ```

"sent" is sort of confusing - this is actually "enqueued", right?

---

*[storage/store.go, line 2206 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6X5laPK40WFynNs43:-KQ6X5laPK40WFynNs44:bnomo0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store.go#L2206)):*
> ```Go
> 	addedPlaceholder := false
> 
> 	s.metrics.raftRcvdMessages[req.Message.Type].Inc(1)
> ```

s/Rcvd/Received/g

None of the other metric names are abbreviated.

---

*[storage/store.go, line 2519 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6X__UEvNdycuvv81P:-KQ6X__UEvNdycuvv81Q:b4x26f2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store.go#L2519)):*
> ```Go
> 			// parallel.
> 			s.metrics.raftLeaders.Clear()
> 			s.metrics.raftLeaseHolders.Clear()
> ```

"raft lease" isn't a terminology we use - can you make it "range lease" throughout?

---

*[storage/store.go, line 2520 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6XwkYslAcBd-6Aq8K:-KQ6XwkYslAcBd-6Aq8L:bk2llsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store.go#L2520)):*
> ```Go
> 			s.metrics.raftLeaders.Clear()
> 			s.metrics.raftLeaseHolders.Clear()
> 			s.metrics.raftLeaseHoldersWithoutLeadership.Clear()
> ```

"rangeLeaseHoldersWithoutRaftLeadership"?

---

*[storage/store_metrics.go, line 272 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4oknM-lR1vjvupioS:-KQ6YWsCVz9DeHS1DKK7:bii68eo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L272)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is a weird pattern. Why is the accumulator a `*metric.Counter` instead of a `*int64`?
</details>
(or just `int64`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 5 files at r3.
Review status: all files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1539 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6WBZoLOSfXOCi-AF7:-KQ6WBZoLOSfXOCi-AF8:b4fz8r6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1539)):*
> ```Go
> }
> 
> func (r *Replica) isLeaseHolder() bool {
> ```

No indication here that this needs a lock.

---

*[storage/store.go, line 2554 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6Xi7grovFLBOptrmq:-KQ6Xi7hF230TFStZNwy:b-u9uw7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store.go#L2554)):*
> ```Go
> 			s.metrics.RaftLeaders.Update(s.metrics.raftLeaders.Count())
> 			s.metrics.RaftLeaseHolders.Update(s.metrics.raftLeaseHolders.Count())
> 			s.metrics.RaftLeaseHoldersWithoutLeadership.Update(s.metrics.raftLeaseHoldersWithoutLeadership.Count())
> ```

Does this really count right? I might understand the code wrong, but it looks like `raftLeaseHoldersWithoutLeadership` is a counter that you clear every time the Store gets ready to call `handleRaftReady`, and that after calling you update the gauge with? That seems that this number would just fluctuate between 0 and the number of replicas last processed in parallel, which I'm not sure is any kind of interesting number. I think at the end of it this is probably just Ben's comment above.

---

*[storage/store_metrics.go, line 168 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6Z8BB9qFOvgtOqcmd:-KQ6Z8BB9qFOvgtOqcme:b-xt49tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L168)):*
> ```Go
> 	metaRaftLeaseHolders = metric.Metadata{
> 		Name: "raft.leaseholders",
> 		Help: "Total number of Ranges on this store that are leaseholders.",
> ```

s/Ranges/Replicas/ throughout

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[storage/store_metrics.go, line 272 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4oknM-lR1vjvupioS:-KQ6ZRU65BFs2mCFhKuX:btyritc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L272)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> (or just `int64`)
</details>
It would need to be behind a mutex lock due to the concurrent updates by different replicas. Should I use `atomic.Value`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[storage/store_metrics.go, line 272 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4oknM-lR1vjvupioS:-KQ6ZlFwCPfA_Gn7EEj_:b-pscr1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L272)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Ravi Narayan) wrote…</i></summary>
> It would need to be behind a mutex lock due to the concurrent updates by different replicas. Should I use `atomic.Value`?
</details>
I believe a bunch of these get updated together, so a mutex would be OK with me. Otherwise, use an `int64` and update it using https://godoc.org/sync/atomic#AddInt64.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions, all commit checks successful.

---

*[storage/store_metrics.go, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6_5vFgYlm68Qc5vmS:-KQ6_5vFgYlm68Qc5vmT:bp4zntf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L279)):*
> ```Go
> 
> 	// Stats for efficient merges.
> 	// TODO(mrtracy): This should be removed as part of #4465. This is only
> ```

@arjunravinarayan this comment appears to be misplaced as a result of your changes here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 22 unresolved discussions, all commit checks successful.

---

*[storage/store_metrics.go, line 272 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4oknM-lR1vjvupioS:-KQ6jtdJVkS3tmC1bx1m:b-qtqpd0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L272)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I believe a bunch of these get updated together, so a mutex would be OK with me. Otherwise, use an `int64` and update it using https://godoc.org/sync/atomic#AddInt64.
</details>
An atomically updated int64 sounds good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 0 of 5 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1539 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6WBZoLOSfXOCi-AF7:-KQ6ZpBs-OvyMGikAAVE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1539)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> No indication here that this needs a lock.
</details>
Done.

---

*[storage/replica.go, line 1563 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4nAVDuqPbRvKl5yyP:-KQ6iC6PPZNV0KVo7gzV:btpw0lv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1563)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> The counter is being reset just before the iteration over replicas + call to `handleRaftReady` - what skew are you referring to?
</details>
The skew Ben is referring to is the fact that `handleRaftReady` is the wrong place to do this - we should really be doing this per tick. I have moved this to the location where `r.tick()` is called over all replicas. I have also put it in its own separate loop where new functions - `isRangeLeaseHolderLocked` and `isRaftLeaderLocked` are called. This obviates the need for atomically incremented counters, and now I just need to update a local variable. It's much cleaner now, PTAL.

---

*[storage/replica.go, line 1902 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6WO5LUIsPN9nBMQ1b:-KQ6_fkFM_bPWvK8oWG7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1902)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> don't you want to grab all the queues?
> 
> ```
> var queuedMsgs int64
> for _, transportQueues := range r.store.ctx.Transport.mu.queues {
>   for _, queue := range transportQueues {
>     queuedMsgs += int64(len(queue))
>   }
> }
> ```
</details>
Done.

---

*[storage/replica.go, line 1909 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6WjU6qbOo9uKSRnsd:-KQ6i5ZhM-Yh4ytxc0HD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/replica.go#L1909)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "sent" is sort of confusing - this is actually "enqueued", right?
</details>
Done.

---

*[storage/store.go, line 2206 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6X5laPK40WFynNs43:-KQ6amS3nXWlwOcXs_ED:bcbafcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store.go#L2206)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Rcvd/Received/g
> 
> None of the other metric names are abbreviated.
</details>
Switched everything to the abbreviated format consistently.

---

*[storage/store.go, line 2519 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6X__UEvNdycuvv81P:-KQ6i7PmmOLgItP9wux6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store.go#L2519)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "raft lease" isn't a terminology we use - can you make it "range lease" throughout?
</details>
Done.

---

*[storage/store.go, line 2520 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6XwkYslAcBd-6Aq8K:-KQ6aqGdj9EY5Pm-stZ1:b-yhculc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store.go#L2520)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "rangeLeaseHoldersWithoutRaftLeadership"?
</details>
Much better name! Done.

---

*[storage/store.go, line 2554 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6Xi7grovFLBOptrmq:-KQ6k7bERYC3DrD1HsHf:b-j68sg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store.go#L2554)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Does this really count right? I might understand the code wrong, but it looks like `raftLeaseHoldersWithoutLeadership` is a counter that you clear every time the Store gets ready to call `handleRaftReady`, and that after calling you update the gauge with? That seems that this number would just fluctuate between 0 and the number of replicas last processed in parallel, which I'm not sure is any kind of interesting number. I think at the end of it this is probably just Ben's comment above.
</details>
Moved to around `r.tick()` directly in `store.go`. This also obviates the need for the counters/atomically updated `int64`s.

---

*[storage/store_metrics.go, line 86 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4oMOHSmMxzBSmdUQ0:-KQ6cIHa9ibSGrGMZoLy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L86)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The messages defined as "local" in https://github.com/coreos/etcd/blob/417b9469aa0f3708829f80dfc3325a576ed43ab0/raft/util.go#L49 never pass through the transfer so we'll never have non-zero values for them. These could be excluded from the metrics. 
</details>
Done.

---

*[storage/store_metrics.go, line 168 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6Z8BB9qFOvgtOqcmd:-KQ6e7_uFnG3c4pTTuyQ:b-ml79yu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L168)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/Ranges/Replicas/ throughout
</details>
I've instead changed them all consistently to be `range` instead (at least for leaseholders), consistent with Tamir's recommendation.

---

*[storage/store_metrics.go, line 272 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ4oknM-lR1vjvupioS:-KQ6h5S8pXZI3CpTtsaO:b-lmd9kn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L272)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> An atomically updated int64 sounds good.
</details>
`AddInt64`on an `int64` seems to be the best choice. However, I have now refactored the calls to use just a local variable, so atomically incremented counters are not needed. Thank you in any case! Now I know how to do this more elegantly in the future.

---

*[storage/store_metrics.go, line 279 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6_5vFgYlm68Qc5vmS:-KQ6hVh0UMDhCdPj_C3S:b707mw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L279)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> @arjunravinarayan this comment appears to be misplaced as a result of your changes here.
</details>
Upon further inspection, it seems misplaced from even before my changes. I've moved it to what looks like the likely intended location, PTAL. Also, #4465 was closed in March, perhaps this comment should be reinspected if it is still relevant. cc @mrtracy.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[storage/store_metrics.go, line 168 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6Z8BB9qFOvgtOqcmd:-KQ6vEbGPVzMWPakVMOL:b-6drln3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L168)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> I've instead changed them all consistently to be `range` instead (at least for leaseholders), consistent with Tamir's recommendation.
</details>
In this case Tobi is right. The lease is a range least, but the holder is a replica.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1903 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6vqOg7ELJu5pE74tx:-KQ6vqOg7ELJu5pE74ty:b-2q8zyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/replica.go#L1903)):*
> ```Go
> 	r.store.ctx.Transport.mu.Unlock()
> 	r.store.metrics.RaftEnqueuedPending.Update(queuedMsgs)
> 	r.store.metrics.RaftEnqueuedMessages.Inc(1)
> ```

can we replace this with a `RaftSentMessages`? You'll need to pass it to the transport and increment it there. In fact, the current handling of `RaftDroppedMessages` is also incorrect, because it doesn't account for messages dropped on connection failure whereupon the entire queue is discarded.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1903 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6vqOg7ELJu5pE74tx:-KQ6xEKzrnHja-UIsPEL:b-c18g94) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/replica.go#L1903)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can we replace this with a `RaftSentMessages`? You'll need to pass it to the transport and increment it there. In fact, the current handling of `RaftDroppedMessages` is also incorrect, because it doesn't account for messages dropped on connection failure whereupon the entire queue is discarded.
</details>
I think we should delegate that to a follow-up.

---

*[storage/store.go, line 2586 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6w98ByxZtEOplRPxG:-KQ6w98ByxZtEOplRPxH:bboq0ce) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/store.go#L2586)):*
> ```Go
> 
> 				for _, r := range s.mu.replicas {
> 					r.mu.Lock()
> ```

Now you acquire the lock here twice (`tick` already acquires this). I think the best way to avoid this is to change the signature:

```go
type tickInfo struct {
  Exists bool
  IsRangeLeaseHolder bool
  // ...
}

func (*Replica) tick(info *tickInfo) error {
  // ...
  info.Exists = true
  // ...
}
```

and here you would do

```go
var info tickInfo
for /* ... */ {
  replica.tick(&info)
  // do stuff with current values of `info`.
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2586 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6w98ByxZtEOplRPxG:-KQ6z2xCPvYYdD5_jkDE:b-8u825v) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/store.go#L2586)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Now you acquire the lock here twice (`tick` already acquires this). I think the best way to avoid this is to change the signature:
> 
> ```go
> type tickInfo struct {
>   Exists bool
>   IsRangeLeaseHolder bool
>   // ...
> }
> 
> func (*Replica) tick(info *tickInfo) error {
>   // ...
>   info.Exists = true
>   // ...
> }
> ```
> 
> and here you would do
> 
> ```go
> var info tickInfo
> for /* ... */ {
>   replica.tick(&info)
>   // do stuff with current values of `info`.
> }
> ```
</details>
you could also have tick return the relevant things, obviating the need for an out-parameter.

or make it `tickLocked`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1903 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6vqOg7ELJu5pE74tx:-KQ6zJEB9NmY_Tbol4NB:b-zflkwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/replica.go#L1903)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I think we should delegate that to a follow-up.
</details>
Sure. If we do delegate that, it may be better to avoid this stuff for now since it is (a) misleading and (b) going to result in a nasty merge conflict with streaming snapshots.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1903 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6vqOg7ELJu5pE74tx:-KQ72iuUYb42vZYj8Ro2:b2p4w8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/replica.go#L1903)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Sure. If we do delegate that, it may be better to avoid this stuff for now since it is (a) misleading and (b) going to result in a nasty merge conflict with streaming snapshots.
</details>
Yeah, I'm not opposed to ripping out whatever isn't 100% polished and putting that in a follow-up PR. We should really get the basic version in ASAP and then bike-shed the rest separately.

---

*[storage/store.go, line 2586 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6w98ByxZtEOplRPxG:-KQ71BYULLB6aYuSQ1Wy:b-mhnc7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/store.go#L2586)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you could also have tick return the relevant things, obviating the need for an out-parameter.
> 
> or make it `tickLocked`.
</details>
I suggested the above option because it's only one heap allocation every 200ms and returning a bunch of function values which are mostly bools doesn't read well.
Not crazy about `tickLocked` since we shouldn't lock the replica from the outside at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2586 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6w98ByxZtEOplRPxG:-KQ7398Y5m04svSozbJu:b-ya1yps) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/store.go#L2586)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I suggested the above option because it's only one heap allocation every 200ms and returning a bunch of function values which are mostly bools doesn't read well.
> Not crazy about `tickLocked` since we shouldn't lock the replica from the outside at all.
</details>
I was suggesting that `tick` would return `tickInfo` (rather than a list of random values).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2586 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6w98ByxZtEOplRPxG:-KQ75a4iuM6wZLAC0QxT:b67di43) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/store.go#L2586)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I was suggesting that `tick` would return `tickInfo` (rather than a list of random values).
</details>
Ah, yeah, that would also work. I personally would still prefer to do the one heap allocation since we've learned that putting more on the stack isn't always better (in this case certainly not), though it shouldn't matter at all in this example.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 2 of 6 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1903 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6vqOg7ELJu5pE74tx:-KQ78D1bb3v0EL4qGaON:bxdqajp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/replica.go#L1903)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Yeah, I'm not opposed to ripping out whatever isn't 100% polished and putting that in a follow-up PR. We should really get the basic version in ASAP and then bike-shed the rest separately.
</details>
Ripped out.

---

*[storage/store.go, line 2586 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6w98ByxZtEOplRPxG:-KQ783KW9FnpwdZ_fPXV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/store.go#L2586)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Ah, yeah, that would also work. I personally would still prefer to do the one heap allocation since we've learned that putting more on the stack isn't always better (in this case certainly not), though it shouldn't matter at all in this example.
</details>
Done.

---

*[storage/store_metrics.go, line 168 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6Z8BB9qFOvgtOqcmd:-KQ76S7wME7j2XHTRwQW:b-1pmfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6375bee554f93b48361ae1d42dc697cf50007d29/storage/store_metrics.go#L168)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> In this case Tobi is right. The lease is a range least, but the holder is a replica.
</details>
Ah that makes more sense, thank you. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. PTAL @tamird

---

Reviewed 6 of 6 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1765 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ7AlT9tVOSXyAkp9Te:-KQ7AlTAEXSp6uxIyWib:beprva) ([raw file](https://github.com/cockroachdb/cockroach/blob/698155c26635a7f46ba46ec2d3149432e5d391be/storage/replica.go#L1765)):*
> ```Go
> 
> // tick the Raft group, returning any error and true if the raft group exists
> // and false otherwise.
> ```

Update comment about `info`.

---

*[storage/store.go, line 2586 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ6w98ByxZtEOplRPxG:-KQ7AwI-nVyZ7VwS22Ef:bpevw42) ([raw file](https://github.com/cockroachdb/cockroach/blob/0332b5db5f3e9f467bb3506584b00db79d58e364/storage/store.go#L2586)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Done.
</details>
just for posterity: `(in this case certainly not)` above reads the wrong way: I was pre-iterating that for this example and in this goroutine it really shouldn't matter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-:-KQ7BKBdmlxufGQDTZ5d:b-f8ekvn)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2578 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ7BqqwGOG2_4c8qF5n:-KQ7BqqwGOG2_4c8qF5o:be4v130) ([raw file](https://github.com/cockroachdb/cockroach/blob/698155c26635a7f46ba46ec2d3149432e5d391be/storage/store.go#L2578)):*
> ```Go
> 
> 				for id, r := range s.mu.replicas {
> 					var info tickInfo
> ```

this is an allocation per replica (!)

per offline discussion this can be a `tickAccumulator` which would be simpler and make it harder to "misuse"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 5 of 6 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2578 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ7BqqwGOG2_4c8qF5n:-KQ7CpzwPiBCkQHCIaEC:b-s4z5t8) ([raw file](https://github.com/cockroachdb/cockroach/blob/698155c26635a7f46ba46ec2d3149432e5d391be/storage/store.go#L2578)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is an allocation per replica (!)
> 
> per offline discussion this can be a `tickAccumulator` which would be simpler and make it harder to "misuse"
</details>
Right, this is pretty bad. Probably better to go with your suggestion, all other things being equal and yours harder to get wrong. `tickAccumulator` does sound reasonable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 5 of 6 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2578 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ7BqqwGOG2_4c8qF5n:-KQ7D1QcC-zlgJ-nHtxm:ba3n63p) ([raw file](https://github.com/cockroachdb/cockroach/blob/698155c26635a7f46ba46ec2d3149432e5d391be/storage/store.go#L2578)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Right, this is pretty bad. Probably better to go with your suggestion, all other things being equal and yours harder to get wrong. `tickAccumulator` does sound reasonable.
</details>
(but that could be done follow-up).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
*[storage/store.go, line 2578 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ7BqqwGOG2_4c8qF5n:-KQ7DDUVoJwuPLvNHZhw:b-dxif34) ([raw file](https://github.com/cockroachdb/cockroach/blob/698155c26635a7f46ba46ec2d3149432e5d391be/storage/store.go#L2578)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> (but that could be done follow-up).
</details>
Fixed, but without the `tickAccumulator`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Review status: 3 of 6 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1765 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ7AlT9tVOSXyAkp9Te:-KQ7Ev-rwOGRLQnOvNCT:b-y9046x) ([raw file](https://github.com/cockroachdb/cockroach/blob/698155c26635a7f46ba46ec2d3149432e5d391be/storage/replica.go#L1765)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Update comment about `info`.
</details>
Updated, and changed the return type to just an `error`. Updated `store.go` to use `tickInfo.Exists` as wel.

---

*[storage/store.go, line 2578 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ7BqqwGOG2_4c8qF5n:-KQ7FIH-x_ZteVjtJI4X:bhwa0wd) ([raw file](https://github.com/cockroachdb/cockroach/blob/698155c26635a7f46ba46ec2d3149432e5d391be/storage/store.go#L2578)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Fixed, but without the `tickAccumulator`.
</details>
Does `tickAccumulator` still make sense if we are holdings things in the struct that aren't accumulated? Like `Exists`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 3 of 6 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2578 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-KQ7BqqwGOG2_4c8qF5n:-KQ7GBRYh9LSWUdA6Ne9:b-ecr4zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/698155c26635a7f46ba46ec2d3149432e5d391be/storage/store.go#L2578)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Does `tickAccumulator` still make sense if we are holdings things in the struct that aren't accumulated? Like `Exists`?
</details>
I was imagining that this part of the API would remain unchanged (that is, the return type would be `(bool, error)` as before).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Anything left to discuss here? The accumulator discussion should be follow-up.

---

Reviewed 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803#-:-KQ7GmL6JJIM-QkXfWHU:bf8y9nu)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8803)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Merging as requested by @arjunravinarayan, teamcity failure appears to be related to failed `git clean`.
